elasticsearch-details/impersonation.md

```diff
@@ -0,0 +1,107 @@
+
+So, an impersonation can be understood as imitating behaviors or actions.
+In the context of ReadonlyREST eg. one user could imitate an action 
+of another user. Why would we want it? Let's suppose the first user is 
+an admin, who has just configured access for a new user. He would like 
+to know if the rule or rules are configured correctly. And here it comes the impersonation feature. The admin can call ES request on behalf of the 
+tested user and authenticate using his credentials. 
+
+## Impersonators configuration
+
+Not anyone can be the impersonator. The list of impersonators and their permissions are defined in `impersonation` section:
+
+```yaml
+readonlyrest:
+  
+  access_control_rules:
+
+    [...]
+
+  impersonation:
+
+    - impersonator: admin1      // impersonator user name or pattern
+      users: ["*"]              // user name patterns that can be impersonated by the impersonator
+      auth_key: admin1:pass     // authentication rule (any authentication rule can be used here)
+
+    - impersonator: admin2
+      users: ["dev2"]
+      ldap_authentication: "ldap1"
+```
+
+We can read it as follows:
+
+1. The impersonator "admin1" can impersonate users which names match pattern `*` (any user) and the impersonator should be authenticated using `auth_key` rule.
+
+2. The impersonator "admin2" can impersonate only `dev2` and the impersonator should be authenticated using LDAP by `ldap1` connector.
+
+## Request with user impersonation
+
+Let's suppose our `user1` wants to search `twitter` index. He probably will call:
+
+```bash
+curl -vk -u user1:user_secret "https://localhost:9200/twitter/_search?pretty"
+```
+
+Now, how the same request would look like it our `admin1` wants to impersonate `user1` and search `twitter` on this behalf?
+
+```bash
+curl -vk -u admin1:pass -H "impersonate_as: user1" "https://localhost:9200/twitter/_search?pretty"
+```
+
+As we can see the request looks pretty much the same. The main difference is the `impersonate_as` header that contains `user1` and authorization header value has the admin's credentials now. 
+
+Responses of both requests should be the same.
+
+When the admin will pass wrong credentials or cannot impersonate the passed user
+ROR is going to return response like this:
+
+> ```text
+> HTTP/1.1 403 Forbidden
+> content-type: application/json; charset=UTF-8
+> 
+>{
+>  "error":{
+>    "root_cause":[
+>      {
+>        "type":"forbidden_response",
+>        "reason":"forbidden",
+>        "due_to":["OPERATION_NOT_ALLOWED", "IMPERSONATION_NOT_ALLOWED"]
+>      }
+>    ],
+>    "type":"forbidden_response",
+>    "reason":"forbidden",
+>    "due_to":["OPERATION_NOT_ALLOWED", "IMPERSONATION_NOT_ALLOWED"]
+>  },
+>  "status":403
+>}
+>```
+
+Some authentication or/and authorization rules (like LDAP ones or related to external systems) don't support impersonation by default (the support will be possible with cooperation with Kibana plugin soon). When request with impersonation
```

Comment:
I don't understand what "by default" means here. And I don't understand how the support for external auth rules will be enabled by the Kibana plugin in the future.

Reply:
You know, impersonation can be easily supported by the rules that have static users and/or groups. If a rule depends on an external system, like an LDAP, you have two ways:
1) depend on impersonation supported by the system
2) mock response from the system

We picked the 2nd way. There is ROR Auth Mock API that can be used to configure the mocked response from LDAP. Eg. ldap1 has users u1 and u2 with groups: [g1, g2] and [g1, g3]. The API at this moment should be considered as an internal one - that's why it's described only in ROR internal docs. Kibana could take advantage of the API (the flow was agreed with the customer) and configure mock responses for the LDAP connector for the tests with impesonation. 

To sum up:
eg. `ldap_auth` rule doesn't support impersonation by default, but then you configure, using the internal Auth Mock API, the LDAP connector mocked users and groups, the impersonation will be supported by the rule.

Reply:
OK So because the configuration is supported via mock API only, and because mock API is internal, we tell the user wait for our Kibana UI if you want to impersonate. 

Do we have a rough idea of how this UI should visually look like? Is there any  similar UI in any other software we can get familiar with and  take as an example?

Reply:
there is a Jira for the topic. Let's move the discussion there. RORDEV-484